### PR TITLE
tr1/text: improve text handling in TR1

### LINKED ISSUE #1919: Accents in Spanish words
Good afternoon, the accents in the Spanish words have disappeared, I attach two images, one that works in the final version 4.5.1 and one that does not work in version 4.6.
How do you play at the demo level?
Thank you very much.

### LINKED ISSUE #1928: Some symbols are not displayed in the game
Hello! Thank you for your hard and interesting work!
I am translating TR1X into Russian, and using a modified texture of standard symbols.
In this texture, the symbols "<" and "," are replaced with Russian "ы" and "ь". This worked well before, but in TR1X-4.6 these symbols are not displayed in the game. I tried to escape them in different ways, but with no luck.
How can I fix this without redrawing the texture? Thank you!)
2024-11-20_100324_cr

### PR DESCRIPTION
**Pull request by rr-**
#### Checklist

- [X] I have read the [coding conventions](https://github.com/LostArtefacts/TR1X/blob/master/CONTRIBUTING.md#coding-conventions)
- [X] I have added a changelog entry about what my pull request accomplishes, or it is an internal change

#### Description

This pull request extends the named sequences support we have introduced in 4.6 to also support Unicode strings. I have confirmed the following languages to have full coverage:

| | |
| ----       | ---- |
| Basque     | Irish               |
| Belarusian | Italian             |
| Bosnian    | Latvian             |
| Bulgarian  | Lithuanian          |
| Catalan    | Macedonian          |
| Croatian   | Malay               |
| Czech      | Maltese             |
| Danish,    | Northern Sami       |
| Dutch      | Norwegian           |
| English    | Polish,             |
| Estonian   | Portuguese          |
| Faroese    | Romanian            |
| Finnish    | Russian             |
| French     | Serbian             |
| Galician   | Slovak              |
| German,    | Slovenian           |
| Greek      | Spanish             |
| Hungarian  | Swedish             |
| Icelandic  | Turkish             |
| Indonesian | …and possibly more. |

Asian and Arabic languages remain unsupported at the moment. While Arabic is still far away due to the RTL rendering order, we're _much_ closer to supporting CJK.
The sprites for the characters come from Arsunt's extended font posted in Tomb Raider Forums.

Pivotal for this feature is a textfile containing manual Unicode codepoint mappings. Although I initially experimented with JSON, YAML, and CSV, I discovered through testing that using a DSL (domain-specific language) designed specifically for this purpose offers the best readability.
The mapping file is used by the tooling in the tools/glyphs/ directory and serves two roles:

1. Hardcoding Unicode to sprite mapping
    It generates C macros that map Unicode code points and escaped sequences to O_ALPHABET's sprite indices, specify glyph dimensions, and instruct how to compose compound characters - all getting hardcoded into the executable.
2. Guidance for font.bin creation
    It directs the injector tool in creating the font.bin file that contains O_ALPHABET sprite bitmaps, along with additional positional information.

Some sprite indices are fixed. This is for compatibility with the original game to retains original text format even if font.bin goes missing.
Creating sprites for all possible accented characters is a challenging and resource-intensive task. Instead, the mapping allows us to combine certain characters so that the game overlays one glyph on another. However, we only support one accent per glyph. Consequently, Vietnamese, despite using the Latin alphabet, is currently unsupported due to its extensive use of diacritics.

As we now have many more glyphs to compare, the time-consuming O(n^2) loop that matched user string characters with all possible glyphs has been replaced by `uthash` lookups for faster glyph retrieval. This approach requires precise knowledge of glyph sizes, necessitating some additional parsing, but it benefits from eliminating ambiguity in glyph matches. An additional benefit is improved handling of Unicode codepoints without declared mappings: by traversing entire codepoints rather than incrementing the pointer by 1 byte, the process avoids ending up in the middle of an incomplete UTF-8 codepoint, preventing garbled text.

### CODE REVIEW

**Comment by aredfan** [comment]
Overall LGTM.  The only issue I found is the TRUB gameflow doesn't point to the font.bin file, which causes this issue.

20241124_180430_Return_to_Egypt

**Comment by rr-** [comment]
@aredfan fixed and made sure that we do not attempt to draw unavailable sprites.

File: data/tr1/ship/cfg/TR1X_gameflow.json5 [hunk 1]
```diff
@@ -482,6 +485,7 @@
             "injections": [
                 "data/injections/backpack_cut.bin",
                 "data/injections/braid_cut2_cut4.bin",
+                "data/injections/font.bin",
```
Comment (walkawayy) [inline data/tr1/ship/cfg/TR1X_gameflow.json5:488]: Do cutscenes not benefit from global injections?

**Review by walkawayy — APPROVED** [review]

File: data/tr1/ship/cfg/TR1X_gameflow.json5 (hunk 1 above)
Comment (rr-) [reply data/tr1/ship/cfg/TR1X_gameflow.json5:488]: They have them disabled with `inherit_injections` set to false – most likely due to `O_LARA` not being present.

**Review by aredfan — APPROVED** [review]

**Review by lahm86 — APPROVED** [review]
Looks fantastic, thank you for doing this.
Noticed a small issue with scaling in the details menu, but it's on develop too - will raise a separate ticket.

**Pull request merged**